src/picongpu/include/particles/manipulators/FreeImpl.hpp

```diff
@@ -31,35 +32,138 @@ namespace particles
 namespace manipulators
 {
 
-template<typename T_Functor>
-struct FreeImpl
-{
+    /** generic manipulator to create user defined manipulators
+     *
+     * @tparam T_Functor user defined functor
+     *
+     * `T_Functor`
+     *   - must implement `void operator()(ParticleType)` or `void operator()(ParticleType1, ParticleType2)`
+     *   - `T_Functor` can implement **only** one `operator()` signature
+     *   - can implement **optional** one host side constructor `T_Functor()` or `T_Functor(uint32_t currentTimeStep)`
+     */
+    template< typename T_Functor >
+    struct FreeImpl : private T_Functor
+    {
 
-    typedef T_Functor Functor;
+        typedef T_Functor Functor;
 
-    template<typename T_SpeciesType>
-    struct apply
-    {
-        typedef FreeImpl<T_Functor> type;
-    };
+        template< typename T_SpeciesType >
+        struct apply
+        {
+            typedef FreeImpl< T_Functor > type;
+        };
 
-    HINLINE FreeImpl(uint32_t)
-    {
-    }
 
-    template<typename T_Particle1, typename T_Particle2>
-    DINLINE void operator()(const DataSpace<simDim>&,
-                            T_Particle1& particleSpecies1, T_Particle2& particleSpecies2,
-                            const bool isParticle1, const bool isParticle2)
-    {
-        if (isParticle1 && isParticle2)
+        /** constructor
+         *
+         * is activated if the user functor has a constructor with one (uint32_t) argument
```

Comment:
No this is not for passing the current time step to the device. It passes the current time step to the user host functor.